acvm/src/pwg/mod.rs

```diff
@@ -140,10 +140,17 @@ pub struct ACVM<'backend, B: BlackBoxFunctionSolver> {
     instruction_pointer: usize,
 
     witness_map: WitnessMap,
+
+    assert_messages: BTreeMap<OpcodeLocation, String>,
 }
 
 impl<'backend, B: BlackBoxFunctionSolver> ACVM<'backend, B> {
-    pub fn new(backend: &'backend B, opcodes: Vec<Opcode>, initial_witness: WitnessMap) -> Self {
+    pub fn new(
+        backend: &'backend B,
+        opcodes: Vec<Opcode>,
+        initial_witness: WitnessMap,
+        assert_messages: BTreeMap<OpcodeLocation, String>,
```

Comment:
To me it feels unnatural that the ACVM doesn't take a circuit instead of the vector of opcodes 🤔 
So if the ACVM is not the one that completes the error with the assert message, I guess that nargo/acvm_js would be the one that does?

Reply:
I see this as similar to solidity's custom revert messages and in this case the EVM never gets any information about these messages passed into it. You need the ABI to decode the EVM's output on a revert based on the selector emitted.

We've got a similar setup except that we don't need a selector due to being able to check which opcode failed.

Reply:
Makes sense! I'm pushing a version that pushes this responsability to acvm_js and nargo, the ones that are on top of the acvm and that have access to the parsed circuit